### PR TITLE
#824: fix git url with branch

### LINKED ISSUE #824: ide create «settings-url»#«branch» not working
### Expected behavior

As a IDEasy settings admin, I want to create a IDEasy project specifying a feature branch to my settings git repo so that I can test the changes before developers pull in updates that may break something and block their daily work.

See the feature described in [usage](https://github.com/devonfw/IDEasy/blob/main/documentation/usage.adoc).

### Actual behavior

```
$ ide create test-settings-branch https://github.com/devonfw/ide-settings.git#dev-ide-settings
Creating new IDEasy project in D:\projects\test-settings-branch
Start: Clone settings repository
Cloning into '.'...
fatal: unable to update url base from redirection:
  asked for: https://github.com/devonfw/ide-settings.git#dev-ide-settings/info/refs?service=git-upload-pack
   redirect: https://github.com/devonfw/ide-settings
Running command 'C:\Program Files\Git\cmd\git.exe' with arguments 'clone' '--recursive' 'https://github.com/devonfw/ide-settings.git#dev-ide-settings' '--config' 'core.autocrlf=false' '.'
failed with exit code 128!
Git failed to clone https://github.com/devonfw/ide-settings.git#dev-ide-settings into D:\projects\test-settings-branch\settings.
Successfully updated settings repository.
Templates folder is missing in settings repository.
Start: Install or update software
Successfully installed java in version 23.0.1_11
Start: Install mvn
No templates found in settings git repo neither in D:\projects\test-settings-branch\settings\templates nor in D:\projects\test-settings-branch\settings\devon - configuration broken
Successfully installed mvn in version 3.9.9
Artifact already exists at C:\Users\hohwille\Downloads\ide\default\node-v23.3.0-windows-x64.zip
To force update please delete the file and run again.
Checksum 868cd7af6f0c042944c38b1bc1212d608177478da2a141bd404f9366d8fbd2dd is correct.
Unpacking 100% [===============================================================] 33/33MiB (0:00:03 / 0:00:00) 11.3MiB/s
Successfully installed node in version v23.3.0
Start: Install npm
Successfully installed npm in version 10.9.1
Successfully ended step 'Install or update software'.
Cannot find repositories folder nor projects folder.
Successfully created new project 'test-settings-branch'.
Successfully completed ide (create,test-settings-branch,https://github.com/devonfw/ide-settings.git#dev-ide-settings)
$ cd test-settings-branch/
$ ls -la settings/
total 0
drwxr-xr-x 1 hohwille 1049089 0 Nov 29 18:30 ./
drwxr-xr-x 1 hohwille 1049089 0 Nov 29 18:30 ../
```

### Steps to reproduce (bug) / Use Case of feature request (enhancement)

1. `ide create test-settings-branch https://github.com/devonfw/ide-settings.git#dev-ide-settings`

### Comments/Hints:

```
fatal: unable to update url base from redirection:
  asked for: https://github.com/devonfw/ide-settings.git#dev-ide-settings/info/refs?service=git-upload-pack
```
It seems as if the feature branch was not properly separated from the URL passed to git.
https://github.com/devonfw/IDEasy/blob/d7df00d12f9556e5f2d531e1f3199e2fd068690a/cli/src/main/java/com/devonfw/tools/ide/context/GitContextImpl.java#L139-L168

### Affected version:

* 2024.12.001-beta-SNAPSHOT

### PR DESCRIPTION
**Pull request by hohwille**
fixes #824
* implement parsing of `GitUrl` from string separating URL from branch (was never implemented before)
* completely refactored `GitContext` to use `GitUrl` as parameter
* fixed buggy URL handling (`java.net.URL` cannot handle git ssh url syntax and will always throw an exception)
* improved offline handling (as initially specified we should first try calling git and only do online check if that fails to produce reasonable error message in case we are offline, only if offline mode we need to fail before we try git clone to be consistent).
* improved `ProcessResult`

### CODE REVIEW

**Review by jan-vcapgemini — APPROVED** [review]
LGTM. Co-Reviewed by: @WorkingAmeise
We've added some questions.

File: cli/src/main/java/com/devonfw/tools/ide/git/GitUrl.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  public static final String BRANCH_MASTER = "master";
+
+  /**
+   * The constructor.
+   */
+  public GitUrl {
+    if (url.contains("#")) {
+      String message = "Invalid git URL " + url;
+      assert false : message;
+      System.out.println(message);
```
[the remark this answers is not in the document]
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/git/GitUrl.java:24]: There is no context here and no access to any logger.
I hope this never happens but if it does it is better to get any hint than just doing nothing.
I would also consider raising an exception but maybe some edge-case I cannot think of could lead to a valid git URL without branch itself may contain a hash sign for some crazy reason. With throwing an exception I would create a blocker for such project. With the output the project can still continue and ignore the warning if they think this is correct.
Once we have #404 implemented, we can change this to SLF4J.

The only alternative until then would be to remove this `System.out.println` statement and silently ignore the error if assertions are disabled what will be the case for end-users.

[the remark this answers is not in the document]
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/git/GitUrl.java:24]: OK. Maybe we just remove the statement and when we have #404 we can add it as a valid log statement.

[the remark this answers is not in the document]
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/git/GitUrl.java]: Done.

**Review by jan-vcapgemini — APPROVED** [review]
Thanks, ready for merge.

**Pull request merged**